crates/consensus/beacon/src/engine/hooks/snapshot.rs

```diff
@@ -0,0 +1,157 @@
+//! Snapshot hook for the engine implementation.
+
+use crate::{
+    engine::hooks::{
+        EngineContext, EngineHook, EngineHookAction, EngineHookError, EngineHookEvent,
+    },
+    hooks::EngineHookDBAccessLevel,
+};
+use futures::FutureExt;
+use reth_db::database::Database;
+use reth_interfaces::{RethError, RethResult};
+use reth_primitives::BlockNumber;
+use reth_snapshot::{Snapshotter, SnapshotterError, SnapshotterWithResult};
+use reth_tasks::TaskSpawner;
+use std::task::{ready, Context, Poll};
+use tokio::sync::oneshot;
+
+/// Manages snapshotting under the control of the engine.
+///
+/// This type controls the [Snapshotter].
+#[derive(Debug)]
+pub struct SnapshotHook<DB> {
+    /// The current state of the snapshotter.
+    state: SnapshotterState<DB>,
+    /// The type that can spawn the snapshotter task.
+    task_spawner: Box<dyn TaskSpawner>,
+}
+
+impl<DB: Database + 'static> SnapshotHook<DB> {
+    /// Create a new instance
+    pub fn new(snapshotter: Snapshotter<DB>, task_spawner: Box<dyn TaskSpawner>) -> Self {
+        Self { state: SnapshotterState::Idle(Some(snapshotter)), task_spawner }
+    }
+
+    /// Advances the snapshotter state.
+    ///
+    /// This checks for the result in the channel, or returns pending if the snapshotter is idle.
+    fn poll_snapshotter(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<RethResult<(EngineHookEvent, Option<EngineHookAction>)>> {
+        let result = match self.state {
+            SnapshotterState::Idle(_) => return Poll::Pending,
+            SnapshotterState::Running(ref mut fut) => {
+                ready!(fut.poll_unpin(cx))
+            }
+        };
+
+        let event = match result {
+            Ok((snapshotter, result)) => {
+                self.state = SnapshotterState::Idle(Some(snapshotter));
+
+                match result {
+                    Ok(_) => EngineHookEvent::Finished(Ok(())),
+                    Err(err) => EngineHookEvent::Finished(Err(match err {
+                        SnapshotterError::InconsistentData(_) => {
+                            EngineHookError::Internal(Box::new(err))
+                        }
+                        SnapshotterError::Interface(err) => err.into(),
+                        SnapshotterError::Database(err) => RethError::Database(err).into(),
+                        SnapshotterError::Provider(err) => RethError::Provider(err).into(),
+                    })),
+                }
+            }
```

Comment:
pedantic style nit:
match conversion and `Err(err)` in two steps, dislike `Err(match {...}`

Reply:
added `From` impls